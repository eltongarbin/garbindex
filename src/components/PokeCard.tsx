import React, { MouseEvent } from 'react';
import { Card, IconButton, CardActionArea } from '@material-ui/core';
import {
  Visibility as VisibilityIcon,
  Delete as DeleteIcon,
  AddCircle as AddCircleIcon
} from '@material-ui/icons';

import {
  CardHeaderStyled,
  CardMediaStyled,
  CardActionsStyled
} from './PokeCardStyled';

type CaptureStates = 'Release' | 'Catch';

type PokeCardProps = {
  image: string;
  id: number;
  name: string;
  captureState: {
    nextState: CaptureStates;
    onChange: (event: MouseEvent<HTMLElement>) => void;
  };
  onSeeMoreClick: (event: MouseEvent<HTMLElement>) => void;
};

const PokeCard = (props: PokeCardProps) => (
  <Card data-testid={`pokecard-${props.id}`}>
    <CardActionArea onClick={props.onSeeMoreClick}>
      <CardHeaderStyled
        title={props.name}
        subheader={`#${props.id}`}
        data-testid="pokecard-header"
      />
      <CardMediaStyled
        image={props.image}
        title={props.name}
        data-testid="pokecard-image"
      />
    </CardActionArea>
    <CardActionsStyled>
      <IconButton aria-label="See more" onClick={props.onSeeMoreClick}>
        <VisibilityIcon />
      </IconButton>
      <IconButton
        aria-label={props.captureState.nextState}
        onClick={props.captureState.onChange}
      >
        {props.captureState.nextState === 'Release' && <DeleteIcon />}
        {props.captureState.nextState === 'Catch' && <AddCircleIcon />}
      </IconButton>
    </CardActionsStyled>
  </Card>
);

export default PokeCard;
