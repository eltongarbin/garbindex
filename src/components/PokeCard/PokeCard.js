import React from 'react';
import PropTypes from 'prop-types';
import { Card, IconButton, CardActionArea } from '@material-ui/core';
import {
  Visibility as VisibilityIcon,
  Delete as DeleteIcon,
  AddCircle as AddCircleIcon
} from '@material-ui/icons';

import {
  CardHeaderStyled,
  CardMediaStyled,
  CardActionsStyled
} from './PokeCardStyled';

function PokeCard({
  image,
  id,
  name,
  onSeeMoreClick,
  onReleaseClick,
  onCatchClick
}) {
  return (
    <Card>
      <CardActionArea onClick={onSeeMoreClick}>
        <CardHeaderStyled
          title={name}
          subheader={`#${id}`}
          data-testid="pokecard-header"
        />
        <CardMediaStyled
          image={image}
          title={name}
          data-testid="pokecard-image"
        />
      </CardActionArea>
      <CardActionsStyled>
        <IconButton aria-label="See more" onClick={onSeeMoreClick}>
          <VisibilityIcon />
        </IconButton>
        {onReleaseClick && (
          <IconButton aria-label="Release" onClick={onReleaseClick}>
            <DeleteIcon />
          </IconButton>
        )}
        {onCatchClick && (
          <IconButton aria-label="Catch" onClick={onCatchClick}>
            <AddCircleIcon />
          </IconButton>
        )}
      </CardActionsStyled>
    </Card>
  );
}

PokeCard.propTypes = {
  image: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onSeeMoreClick: PropTypes.func.isRequired,
  onReleaseClick: PropTypes.func,
  onCatchClick: PropTypes.func
};

export default PokeCard;
