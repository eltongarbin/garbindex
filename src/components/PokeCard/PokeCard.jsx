import React from 'react';
import PropTypes from 'prop-types';
import { Card, IconButton, CardActionArea } from '@material-ui/core';
import {
  Visibility as VisibilityIcon,
  Delete as DeleteIcon,
  AddCircle as AddCircleIcon
} from '@material-ui/icons';

import {
  CardHeaderStyled,
  CardMediaStyled,
  CardActionsStyled
} from './PokeCardStyled';

function PokeCard({
  classes,
  image,
  id,
  name,
  onSeeMoreClick,
  onDeleteClick,
  onCatchClick
}) {
  return (
    <Card>
      <CardActionArea onClick={onSeeMoreClick}>
        <CardHeaderStyled title={name} subheader={`#${id}`} />
        <CardMediaStyled image={image} title={name} />
      </CardActionArea>
      <CardActionsStyled>
        <IconButton aria-label="See more" onClick={onSeeMoreClick}>
          <VisibilityIcon />
        </IconButton>
        {onDeleteClick && (
          <IconButton aria-label="Delete" onClick={onDeleteClick}>
            <DeleteIcon />
          </IconButton>
        )}
        {onCatchClick && (
          <IconButton aria-label="Catch" onClick={onCatchClick}>
            <AddCircleIcon />
          </IconButton>
        )}
      </CardActionsStyled>
    </Card>
  );
}

PokeCard.propTypes = {
  image: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onSeeMoreClick: PropTypes.func.isRequired,
  onDeleteClick: PropTypes.func,
  onCatchClick: PropTypes.func
};

export default PokeCard;
