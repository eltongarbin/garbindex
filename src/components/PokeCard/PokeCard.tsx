import React, { MouseEvent } from 'react';
import { Card, IconButton, CardActionArea } from '@material-ui/core';
import {
  Visibility as VisibilityIcon,
  Delete as DeleteIcon,
  AddCircle as AddCircleIcon
} from '@material-ui/icons';

import {
  CardHeaderStyled,
  CardMediaStyled,
  CardActionsStyled
} from './PokeCardStyled';

type PokeCardProps = {
  image: string;
  id: number;
  name: string;
  onSeeMoreClick: (event: MouseEvent<HTMLElement>) => void;
  onReleaseClick?: (event: MouseEvent<HTMLElement>) => void;
  onCatchClick?: (event: MouseEvent<HTMLElement>) => void;
};

const PokeCard = (props: PokeCardProps) => (
  <Card data-testid="pokecard">
    <CardActionArea onClick={props.onSeeMoreClick}>
      <CardHeaderStyled
        title={props.name}
        subheader={`#${props.id}`}
        data-testid="pokecard-header"
      />
      <CardMediaStyled
        image={props.image}
        title={props.name}
        data-testid="pokecard-image"
      />
    </CardActionArea>
    <CardActionsStyled>
      <IconButton aria-label="See more" onClick={props.onSeeMoreClick}>
        <VisibilityIcon />
      </IconButton>
      {props.onReleaseClick && (
        <IconButton aria-label="Release" onClick={props.onReleaseClick}>
          <DeleteIcon />
        </IconButton>
      )}
      {props.onCatchClick && (
        <IconButton aria-label="Catch" onClick={props.onCatchClick}>
          <AddCircleIcon />
        </IconButton>
      )}
    </CardActionsStyled>
  </Card>
);

export default PokeCard;
