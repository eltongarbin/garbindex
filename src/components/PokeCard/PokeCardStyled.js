import styled from 'styled-components';
import { CardHeader, CardMedia, CardActions } from '@material-ui/core';

export const CardHeaderStyled = styled(CardHeader)`
  && {
    padding-bottom: 0px;
  }
`;

export const CardMediaStyled = styled(CardMedia)`
  && {
    background-size: auto;
    height: 100px;
  }
`;

export const CardActionsStyled = styled(CardActions)`
  && {
    display: flex;
    justify-content: space-between;
  }
`;
