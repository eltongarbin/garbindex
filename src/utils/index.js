export { default as history } from './history';
export { default as theme } from './theme';
