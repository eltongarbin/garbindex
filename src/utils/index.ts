export { default as history } from './history';
export { default as theme } from './theme';

export const extractParamId = (url: any) => {
  const [id] = url.match(/(\d+)(?!.*\d)/);
  return id;
};
