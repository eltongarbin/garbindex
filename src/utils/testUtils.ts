/* eslint-disable react/forbid-foreign-prop-types */
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, conformingProps) => {
  const propError = checkPropTypes(
    component.propTypes,
    conformingProps,
    'prop',
    component.name
  );

  expect(propError).toBeUndefined();
};
