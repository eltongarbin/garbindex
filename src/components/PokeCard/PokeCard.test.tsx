import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CardActionArea, IconButton } from '@material-ui/core';
import {
  Delete as DeleteIcon,
  AddCircle as AddCircleIcon
} from '@material-ui/icons';

import { checkProps } from 'utils/testUtils';
import PokeCard from './PokeCard';

const expectedRequiredProps = {
  id: 1,
  image: 'image',
  name: 'name',
  onSeeMoreClick: () => {}
};

describe('<PokeCard />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<PokeCard {...expectedRequiredProps} />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('does not throw warning with expected props', () => {
    checkProps(PokeCard, expectedRequiredProps);
  });

  it('calls `onSeeMoreClick` prop upon button click', () => {
    const buttonActionMock = jest.fn();
    const wrapper = shallow(
      <PokeCard {...expectedRequiredProps} onSeeMoreClick={buttonActionMock} />
    );

    wrapper.find(CardActionArea).simulate('click');
    wrapper.find(IconButton).simulate('click');

    expect(buttonActionMock.mock.calls.length).toBe(2);
  });

  it('should render <DeleteIcon /> when pass `onReleaseClick` prop', () => {
    const wrapper = shallow(
      <PokeCard {...expectedRequiredProps} onReleaseClick={() => {}} />
    );
    expect(wrapper.find(DeleteIcon)).toHaveLength(1);
  });

  it('calls `onReleaseClick` prop upon button click', () => {
    const buttonActionMock = jest.fn();
    const wrapper = shallow(
      <PokeCard {...expectedRequiredProps} onReleaseClick={buttonActionMock} />
    );

    wrapper
      .find(DeleteIcon)
      .parent()
      .simulate('click');

    expect(buttonActionMock.mock.calls.length).toBe(1);
  });

  it('should not <DeleteIcon /> when not pass `onReleaseClick` prop', () => {
    const wrapper = shallow(<PokeCard {...expectedRequiredProps} />);
    expect(wrapper.find(DeleteIcon)).toHaveLength(0);
  });

  it('should render <AddCircleIcon /> when pass `onCatchClick` prop', () => {
    const wrapper = shallow(
      <PokeCard {...expectedRequiredProps} onCatchClick={() => {}} />
    );
    expect(wrapper.find(AddCircleIcon)).toHaveLength(1);
  });

  it('calls `onCatchClick` prop upon button click', () => {
    const buttonActionMock = jest.fn();
    const wrapper = shallow(
      <PokeCard {...expectedRequiredProps} onCatchClick={buttonActionMock} />
    );

    wrapper
      .find(AddCircleIcon)
      .parent()
      .simulate('click');

    expect(buttonActionMock.mock.calls.length).toBe(1);
  });

  it('should not <AddCircleIcon /> when not pass `onCatchClick` prop', () => {
    const wrapper = shallow(<PokeCard {...expectedRequiredProps} />);
    expect(wrapper.find(AddCircleIcon)).toHaveLength(0);
  });
});
