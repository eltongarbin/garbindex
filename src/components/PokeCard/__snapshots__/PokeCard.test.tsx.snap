// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokeCard /> matches the snapshot 1`] = `
<WithStyles(ForwardRef(Card))
  data-testid="pokecard"
>
  <WithStyles(ForwardRef(CardActionArea))
    onClick={[Function]}
  >
    <Styled(WithStyles(ForwardRef(CardHeader)))
      data-testid="pokecard-header"
      subheader="#1"
      title="name"
    />
    <Styled(WithStyles(ForwardRef(CardMedia)))
      data-testid="pokecard-image"
      image="image"
      title="name"
    />
  </WithStyles(ForwardRef(CardActionArea))>
  <Styled(WithStyles(ForwardRef(CardActions)))>
    <WithStyles(ForwardRef(IconButton))
      aria-label="See more"
      onClick={[Function]}
    >
      <VisibilityIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Styled(WithStyles(ForwardRef(CardActions)))>
</WithStyles(ForwardRef(Card))>
`;
